File: frontend/app/src/types/versioning.ts
export type ArtifactStatus = "available" | "missing" | "unknown";

export type VersionRecord = {
  id: string;
  project_id: string;
  version_tag: string;
  commit_hash: string;
  trigger_source: string;
  summary: string;
  plant_graph_path?: string | null;
  logic_path?: string | null;
  io_mapping_path?: string | null;
  simulation_results_path?: string | null;
  runtime_state_path?: string | null;
  created_at: string;
  created_by?: string | null;
  deployment_tag?: string | null;
  rollback_available: boolean;
  artifact_status: {
    plant_graph: ArtifactStatus;
    control_logic: ArtifactStatus;
    io_mapping: ArtifactStatus;
    simulation: ArtifactStatus;
    runtime: ArtifactStatus;
  };
};

export type SnapshotRecord = VersionRecord;

export type VersionDiffResponse = {
  project_id: string;
  version_a: string;
  version_b: string;
  logic_diff: Record<string, string>;
  metadata_diff: Record<string, { from: unknown; to: unknown }>;
};

export type RollbackResponse = {
  project_id: string;
  rolled_back_to: string;
  restored_files: string[];
  rollback_commit: {
    status: string;
    project_id: string;
    version_tag: string;
    commit_hash: string;
    snapshot_path: string;
    trigger_source: string;
    summary?: string;
    artifact_status?: Record<string, string>;
    metadata_id?: string;
  };
};

export type VersionCommitPayload = {
  project_id: string;
  trigger_source: string;
  summary?: string;
};

export type VersionSnapshotPayload = VersionCommitPayload;

export type VersionRollbackPayload = {
  project_id: string;
  version_tag: string;
};

File: backend/models/versioning.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class VersionCommitRequest(BaseModel):
    project_id: str
    trigger_source: str
    summary: str | None = None


class VersionSnapshotRequest(BaseModel):
    project_id: str
    trigger_source: str
    summary: str | None = None


class VersionRollbackRequest(BaseModel):
    project_id: str
    version_tag: str


class VersionDiffQuery(BaseModel):
    compare_to: str | None = None


class VersionActionResponse(BaseModel):
    status: str
    project_id: str
    version_tag: str
    commit_hash: str | None = None
    snapshot_path: str | None = None
    trigger_source: str | None = None
    summary: str | None = None
    artifact_status: dict[str, str] = Field(default_factory=dict)
    metadata_id: str | None = None
    dolt: dict[str, Any] = Field(default_factory=dict)


class VersionHistoryResponse(BaseModel):
    project_id: str
    records: list[dict[str, Any]] = Field(default_factory=list)


class VersionDiffResponse(BaseModel):
    project_id: str
    version_a: str
    version_b: str
    logic_diff: dict[str, str] = Field(default_factory=dict)
    metadata_diff: dict[str, Any] = Field(default_factory=dict)
